### PR TITLE
Fix timeout issues

### PR DESCRIPTION
**Pull request by e111077**
Our test throttling for Safari had to be extended to Chrome. I also changed out throttling implementation:

before we would just run it after certain test suites. Now, I patched `replaceState` to log how many calls it has had. Before and after every test I check to see if it was called more than 85 times since the last timeout by putting the new cooldownFunction in the setup and teardown of each suite. If it has been called more than 90 times since the last timeout, then it times out for 30 seconds and resets the count.

For the patch and new cooldown function [see](https://github.com/PolymerElements/iron-location/pull/109/files#diff-22b5a4c9942bde064aff5294c86fe075R1).

For some high velocity tests, I force the timeout [see](https://github.com/PolymerElements/iron-location/pull/109/files#diff-9c2bf7820c8af6945a0c10b98a7299b4R203)

For everything else, see where I inserted `cooldownFunction`

There unfortunately was some autoformatting happening.

I'm super sorry for the messy PR. GH won't update the diff if I try to merge it in a branch with everything already formatted

### CODE REVIEW

**Comment by e111077** [comment]
updated author

**Comment by e111077** [comment]
Failure error is:
`Your test errored. You've exceeded your Sauce Labs concurrency limit. This test was throttled and ultimately timed out waiting for a free slot to run. You can upgrade your account for higher-concurrency at https://saucelabs.com/pricing`

File: test/replace-state-patch.html [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  var isChrome = /Chrome/.test(navigator.userAgent) && /Google Inc/.test(navigator.vendor);
+  var isSafari = /^Apple/.test(navigator.vendor);
+
+  if (isChrome || isSafari) {
+    cooldownFunction = function (done, force) {
+      var resolve;
+      // this is for setups that return promises
+      var cooledDown = new Promise(function (res) { resolve = res });
```
Comment (rictic) [inline test/replace-state-patch.html]: Here you're doing both Promise-based logic and callback based logic. Why not write in Promise style, and then have a check at the end like:

```javascript
if (done) {
  cooledDown.then(function() {done();}, function (err) {done(err);});
}
```

File: test/replace-state-patch.html [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+      console.log(__runnerWindow.replaceStateCount)
```
Comment (rictic) [inline test/replace-state-patch.html]: Remove debug console.log?

File: test/replace-state-patch.html [hunk 3]
```diff
@@ -0,0 +1,81 @@
+<script>
```
Comment (bicknellr) [inline test/replace-state-patch.html]: This script seems like a good candidate for wrapping in an IIFE; no implicit globals and you could remove some of the `__` prefixing.

Comment (bicknellr) [reply test/replace-state-patch.html]: Is the only thing you intend to 'export' `cooldownFunction`?

File: test/replace-state-patch.html (hunk 1 above)
Comment (e111077) [reply test/replace-state-patch.html]: 👍 👍 👍

File: test/iron-location.html [hunk 4]
```diff
@@ -18,520 +19,531 @@
...
 
-      test('of the query portion of the URL', function() {
-        expect(window.location.search).to.not.be.equal('?redirectedTo');
-        var redirector = fixture('RedirectQuery');
-        expect(window.location.search).to.be.equal('?redirectedTo');
-        expect(redirector.query).to.be.equal('redirectedTo');
-        redirector.query = 'newQuery';
-        expect(window.location.search).to.be.equal('?redirectedTo');
-        expect(redirector.query).to.be.equal('redirectedTo');
+        suite('with a base URI', function () {
```
[the remark this answers is not in the document]
Comment (e111077) [reply test/iron-location.html]: It was initially moved out because duplicating the file temporary fixed the throttling exception

**Comment by e111077** [comment]
Ready for another pass. Test failures are the same. It seems as if tests are taking too long and travis is trying to run too many concurrent browsers as a result

File: test/replace-state-patch.html (hunk 3 above)
Comment (e111077) [reply test/replace-state-patch.html]: implemented

File: test/replace-state-patch.html [hunk 5]
```diff
@@ -46,7 +48,7 @@
         var resolve;
         // this is for setups that return promises
         var cooledDown = new Promise(function (res) {
```
Comment (rictic) [inline test/replace-state-patch.html]: Having cooledDown and resolve and res seems a bit overcomplicated. Why not just:

```javascript
window.cooldownFunction = function (done, force) {
  return new Promise(function (resolve) {
    // Rest of logic in here
  });
};
```

This means more indentation, but you can recover that by doing an early exit if you don't need to do the cooldown. So:

```javascript
var shouldCooldown = runnerWindow.replaceStateCount > 90 || force;
if (!shouldCooldown) {
  resolve();
  return;
}
var cooldownPeriod = 30 * 1000;
// etc
```

**Pull request merged**